swift/ql/lib/codeql/swift/elements/decl/Decl.qll

```diff
@@ -1,4 +1,23 @@
-// generated by codegen/codegen.py, remove this comment if you wish to edit this file
 private import codeql.swift.generated.decl.Decl
+private import codeql.swift.elements.decl.DeclWithMembers
+private import codeql.swift.elements.decl.NominalTypeDecl
+private import codeql.swift.elements.decl.ExtensionDecl
 
-class Decl extends Generated::Decl { }
+class Decl extends Generated::Decl {
+  /**
+   * Gets the class, struct, enum, protocol, or extension that declared
+   * this `Decl` as a member.
+   */
+  cached
```

Comment:
Is there a rule of thumb for when I should cache a predicate?

Reply:
It's always a bit difficult to judge when (and how) something should be cached. Generally, if a predicate is (1) expensive to compute, and (2) used in multiple queries, then it should most likely be cached. In this case, I don't think that the predicate should be cached as I don't think `getDeclaringDecl` is expensive to compute.

Reply:
The simplest rule of thumb is: if in doubt, don't mark predicates  `cached` when you write them (or use `pragma[only_bind_into]` and other hints).  The optimizer doesn't _usually_ need these tags to do a good job, and you can make things worse by using them.  In the case of `cached`, I believe it prevents the optimizer from using the context in which the predicate is called to reduce its size, which can sometimes cost more than any gains.

The time to use these tools IMO is when you've identified a measurable performance problem and you're trying to fix it.  I guess the other situation is when you're implementing or copying a familiar pattern where `cached` has worked out well before.